mypy/build.py

```diff
@@ -1477,6 +1530,25 @@ def process_graph(graph: Graph, manager: BuildManager) -> None:
         else:
             process_stale_scc(graph, scc)
 
+        # TODO: This is a workaround to get around the "chaining imports" problem
+        # with the interface checks.
+        #
+        # That is, if we have a file named `module_a.py` which does:
+        #
+        #     import module_b
+        #     module_b.module_c.foo(3)
+        #
+        # ...and if the type signature of `module_c.foo(...)` were to change,
+        # module_a_ would not be rechecked since the interface of `module_b`
+        # would not be considered changed.
+        #
+        # As a workaround, this check will force a module's interface to be
```

Comment:
With the workaround, the PR makes incremental better then what it's currently doing, but not as good as it could be -- my plan was to break up the work by landing this first then moving on to figuring out how to remove this workaround and submit a second pull request, probably by doing something similar to Jukka's suggestions.

More specifically, suppose we had the following test case:

```python
[case testIncrementalChain]
import a

[file a.py]
import b
def func_a() -> int:
    return b.func_b()
    
[file b.py]
import c
def func_b() -> int:
    return c.func_c()

[file c.py]
def func_c() -> int:
    return 42
```

With the version of incremental mode in master, changing file `c.py` in any way will cause the main module, file `a.py`, `b.py`, and `c.py` to all be rechecked.

This pull request modifies incremental so that if `c.py` was modified without the public interface being changed (maybe changing `func_c` to return a different value?), then only `c.py` is rechecked. However, changing `c.py` so that its interface changes causes mypy to fall back back to its current behavior (rechecking everything). This behaves basically identically to my original (closed) pull request, except the checks are all more robust/I didn't need to resort to writing another visitor.

Ideally, after the workaround is removed, changing the interface of `c.py` should cause only `b.py` and `c.py` to be changed, but that isn't implemented yet and is the next thing I'm planning on working around.